src/block_ffm.rs

```diff
@@ -156,79 +160,118 @@ impl <L:OptimizerTrait + 'static> BlockTrait for BlockFFM<L>
-                                        // Original: wsum += left_hash_weight * right_side;
-                                        *wsumbuf.get_unchecked_mut(k) += left_hash_weight * right_side;
+                                specialize_1f32!(left_hash.value, LEFT_HASH_VALUE, {
+                                    if feature_num == 0 {
+                                        for z in 0..fb.ffm_fields_count {
+                                            _mm_prefetch(mem::transmute::<&f32, &i8>(&ffm_weights.get_unchecked(addr + FFMK as usize).weight), _MM_HINT_T0);
+                                            for k in 0..FFMK as usize{
+                                                *contra_fields.get_unchecked_mut(zfc + k) = ffm_weights.get_unchecked(addr + k).weight * LEFT_HASH_VALUE;
+                                            }
+                                            zfc += fc;
+                                            addr += FFMK as usize
+                                        }
+                                    } else {
+                                        for z in 0..fb.ffm_fields_count {
+                                            _mm_prefetch(mem::transmute::<&f32, &i8>(&ffm_weights.get_unchecked(addr + FFMK as usize).weight), _MM_HINT_T0);
+                                            for k in 0..FFMK as usize{
+                                                *contra_fields.get_unchecked_mut(zfc + k) += ffm_weights.get_unchecked(addr + k).weight * LEFT_HASH_VALUE;
+                                            }
+                                            zfc += fc;
+                                            addr += FFMK as usize
+                                        }
                                     }
                                 });
-                                
+                                ffm_buffer_index += 1;
+                                feature_num += 1;
                             }
-                            ifc += fc;
-                            
                         }
+                        
+                        let mut ffm_values_offset = 0;
+                        for (i, left_hash) in fb.ffm_buffer.iter().enumerate() {
+                            let mut contra_offset = (left_hash.contra_field_index * fb.ffm_fields_count) as usize;
+                            let mut vv = 0;
+                            let left_hash_value = left_hash.value;
+                            let left_hash_contra_field_index = left_hash.contra_field_index;
+                            let left_hash_hash = left_hash.hash as usize;
+                            //let LEFT_HASH_VALUE = left_hash_value;
+                            specialize_1f32!(left_hash_value, LEFT_HASH_VALUE, {
+
+                              for z in 0..fb.ffm_fields_count as usize {
+                                  if vv == left_hash_contra_field_index as usize {
```

Comment:
just making sure I remember correctly - in this case we DO want fields to self interact, but only with other namespaces in the field? (hence the subtraction)